firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/Firestore.kt

```diff
@@ -162,3 +172,67 @@ class FirebaseFirestoreKtxRegistrar : ComponentRegistrar {
     override fun getComponents(): List<Component<*>> =
             listOf(LibraryVersionComponent.create(LIBRARY_NAME, BuildConfig.VERSION_NAME))
 }
+
+/**
+ * Transforms a [DocumentReference] into a coroutine [Flow]
+ *
+ * **Backpressure handling**: by default this method conflates items. If the consumer isn't fast enough,
+ * it might miss some values but is always guaranteed to get the latest value. Use [bufferCapacity]
+ * to change that behaviour
+ *
+ * @param metadataChanges controls metadata-only changes. Default: [MetadataChanges.EXCLUDE]
+ * @param bufferCapacity the buffer capacity as in [Flow.buffer] or null to not buffer at all
+ */
+fun DocumentReference.toFlow(
+    metadataChanges: MetadataChanges = MetadataChanges.EXCLUDE,
+    bufferCapacity: Int? = Channel.CONFLATED
+): Flow<DocumentSnapshot> {
+    val flow = callbackFlow {
+        val registration = addSnapshotListener(metadataChanges) { snapshot, exception ->
+            if (exception != null) {
+                cancel(message = "Error getting DocumentReference snapshot", cause = exception)
+            } else if (snapshot != null) {
+                trySend(snapshot)
+            }
+        }
+        awaitClose { registration.remove() }
+    }
+
+    return if (bufferCapacity != null) {
+        flow.buffer(bufferCapacity)
+    } else {
+        flow
+    }
+}
+
+/**
+ * Transforms a [Query] into a coroutine [Flow]
+ *
+ * **Backpressure handling**: by default this method conflates items. If the consumer isn't fast enough,
+ * it might miss some values but is always guaranteed to get the latest value. Use [bufferCapacity]
+ * to change that behaviour
+ *
+ * @param metadataChanges controls metadata-only changes. Default: [MetadataChanges.EXCLUDE]
+ * @param bufferCapacity the buffer capacity as in [Flow.buffer] or null to not buffer at all
+ */
```

Comment:
I have a humble suggestion to improve the ref docs here, but it might not be final - I will ask a technical writer to review it.

```suggestion
/**
 * Starts listening to this query with the given options and emits its values via a [Flow].
 *
 * When a collection of the returned flow starts, an EventListener will be attached.
 * When a collection completes, the listener will be removed.
 *
 * Backpressure: by default the returned flow is conflated. If the consumer isn't fast enough,
 * it might miss some values but is always guaranteed to get the latest value. Use the [bufferCapacity] parameter
 * to change that behaviour.
 *
 * @param metadataChanges controls metadata-only changes. Default: [MetadataChanges.EXCLUDE]
 * @param bufferCapacity the buffer capacity as in [Flow.buffer] or null to not buffer at all
 */
```

Reply:
I'd skip mentioning `EventListener` as it's an implementation detail that's not relevant to the Kotlin caller.

Reply:
My idea behind mentioning EventListener is in case a developer is trying to troubleshoot some unexpected behavior on the Flow collection. They can refer to EventListener in our documentation.

Reply:
Works for me 👍 . 

PS: I believe "edits from maintainers" are enabled on this PR? Feel free to push to the branch directly or open a new from it if you want to do changes.